src/Workspaces/Core/Portable/Workspace/Host/DocumentService/IDocumentExcerptService.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Host
+{
+    /// <summary>
+    /// excerpt some part of <see cref="Document"/>
+    /// </summary>
+    internal interface IDocumentExcerptService : IDocumentService
+    {
+        /// <summary>
+        /// return <see cref="ExcerptResult"/> of given <see cref="Document"/> and <see cref="TextSpan"/>
+        /// 
+        /// the result might not be an exact copy of the given source or contains more then given span
+        /// </summary>
+        Task<ExcerptResult?> TryExcerptAsync(Document document, TextSpan span, ExcerptMode mode, CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// this mode shows intention not actual behavior. it is up to implementation how to interpret the intention.
+    /// </summary>
+    internal enum ExcerptMode
+    {
+        SingleLine,
+        Tooltip
+    }
+
+    /// <summary>
+    /// Result of excerpt
+    /// </summary>
+    internal struct ExcerptResult
+    {
+        /// <summary>
+        /// excerpt content
+        /// </summary>
+        public readonly SourceText Content;
+
+        /// <summary>
+        /// span on <see cref="Content"/> that given <see cref="Span"/> got mapped to
+        /// </summary>
+        public readonly TextSpan MappedSpan;
```

Comment:
a lot of these comments need to be fleshed out more.  Given that it's called 'MappedSpan' my intuition is that something was mapped.  But what are the circumstances where this mapping happens?  When is that mapping 'interesting' and not jsut 1:1 with the original span pass in, etc. etc.